great_expectations/dataset/base.py

```diff
@@ -821,10 +839,85 @@ def validate(self, expectations_config=None, catch_exceptions=True, result_forma
+                "evaluation_parameters": evaluation_parameters
+            }
+        else:
+            return {
+                "results": results
+            }
+
+    def get_evaluation_parameter(self, parameter_name, default_value=None):
+        """Get an evaluation parameter value that has been stored in meta.
+        
+        Args:
+            parameter_name (string): The name of the parameter to store.
+            default_value (any): The default value to be returned if the parameter is not found.
+
+        Returns:
+            The current value of the evaluation parameter.
+        """
+        if "meta" in self._expectations_config and \
+                "evaluation_parameters" in self._expectations_config["meta"] and \
+                parameter_name in self._expectations_config['meta']['evaluation_parameters']:
+            return self._expectations_config['meta']['evaluation_parameters'][parameter_name]
+        else:
+            return default_value
+
+    def set_evaluation_parameter(self, parameter_name, parameter_value):
+        """Provide a kwarg value to be stored in the dataset meta object and used to evaluate expectations with missing
+        parameters.
+
+        Args:
+            expectation_name (string): The name of the expectation for which this kwarg should be used
+            kwarg_name (string): The name of the kwarg to be replaced at evaluation time
+            kwarg_value (any): The value to be used
+        """
+        # if not callable(getattr(self, expectation_name, None)):
+        #     raise ValueError("The expectation " + expectation_name + " is not available in this dataset.")
+
+        if 'meta' not in self._expectations_config:
+            self._expectations_config['meta'] = { 'evaluation_parameters': {}}
+
+        if 'evaluation_parameters' not in self._expectations_config['meta']:
+            self._expectations_config['meta']['evaluation_parameters'] = {}
+
+        # if expectation_name in self._expectations_config['meta']:
+        #     self._expectations_config['meta'][expectation_name].update({
+        #             kwarg_name: kwarg_value
+        #         })
+        # else:
+        #     self._expectations_config['meta'][expectation_name] = {
+        #         kwarg_name: kwarg_value
+        #     }
+
+        self._expectations_config['meta']['evaluation_parameters'].update({parameter_name: parameter_value})
+
+    def _build_evaluation_parameters(self, expectation_args, evaluation_parameters):
+        """Build a dictionary of parameters to evaluate, using the provided evaluation_paramters,
+        AND mutate expectation_args by removing any parameter values passed in as temporary values during
+        exploratory work.
+        """
+
+        evaluation_args = copy.deepcopy(expectation_args)
+
+        # Iterate over arguments, and replace $PARAMETER-defined args with their
+        # specified parameters.
+        for key, value in evaluation_args.items():
```

Comment:
Does this structure allow for recursion beyond top-level objects?

Ex: if I have a custom expectation that takes `my_arg={"a":{"b":{"c":{"$PARAMETER" : "some_key"}}}}`, will it find parameter?

Similarly: should we be able to pass nested parameters? `my_arg={"a":{"$PARAMETER" : "top_level_key.2nd_level_key.3rd_level_key"}}`? This could be helpful if parameter objects start to contain lots of information.

I think the first example will be important. Not sure about the second.

Reply:
...also, how will we deal with nesting in lists vs dictionaries?

Reply:
I think I need more specific examples of both of those cases to think through them fully. Short answer is that no, there is no nesting of either kind. I'd suggest that would need to be an extension. Essentially that is just a "partially parameterized" expectation.

Reply:
okay. I'm okay running without nesting for now.

I'm about 80% sure that we're going to want type 1 nesting at some point, but we don't need to include it right at the beginning.

Reply:
You get to implement it then! :) Seriously though, it's of course possible to have the parameterized object be nested, so I'm not sure we're losing a lot now.